### PR TITLE
Add Single.doOnError()

### PR DESCRIPTION
**Pull request by artem-zinnatullin**

### CODE REVIEW

File: src/main/java/rx/Single.java [hunk 1]
```diff
@@ -1789,4 +1790,39 @@ public void onNext(T t) {
         return zip(this, other, zipFunction);
     }
 
+    /**
+     * Modifies the source {@link Single} so that it invokes an action if it calls {@code onError}.
+     * <p>
+     * In case the onError action throws, the downstream will receive a composite exception containing
```
Comment (akarnokd) [inline src/main/java/rx/Single.java]: Could you add an unit test that verifies this?

File: src/main/java/rx/Single.java [hunk 2]
```diff
@@ -1789,4 +1790,39 @@ public void onNext(T t) {
...
+     *  <dt><b>Scheduler:</b></dt>
+     *  <dd>{@code doOnError} does not operate by default on a particular {@link Scheduler}.</dd>
+     * </dl>
+     *
+     * @param onError
+     *            the action to invoke if the source {@link Single} calls {@code onError}
+     * @return the source {@link Single} with the side-effecting behavior applied
+     * @see <a href="http://reactivex.io/documentation/operators/do.html">ReactiveX operators documentation: Do</a>
+     */
+    public final Single<T> doOnError(final Action1<Throwable> onError) {
```
Comment (akarnokd) [inline src/main/java/rx/Single.java]: This should start out as experimental unless the RxJava contributors want to fast-track this.

File: src/main/java/rx/Single.java (hunk 1 above)
Comment (artem-zinnatullin) [inline src/main/java/rx/Single.java]: Added `doOnErrorShouldThrowCompositeExceptionIfOnErrorActionThrows()`

File: src/main/java/rx/Single.java (hunk 2 above)
Comment (artem-zinnatullin) [inline src/main/java/rx/Single.java]: Added `@Experimental`

**Comment by akarnokd** [comment]
:+1:

**Comment by abersnaze** [comment]
:+1:

**Pull request merged**